security/pkg/stsservice/mock/xdsserver.go

```diff
@@ -0,0 +1,235 @@
+	server := xds.NewServer(context.Background(), snapshotCache, cb)
+	var gRPCServer *grpc.Server
+	if isTLS {
+		tlsCred, err := credentials.NewServerTLSFromFile(conf.CertFile, conf.KeyFile)
+		if err != nil {
+			xdsServerLog.Errorf("Failed to setup TLS: %v", err)
+			return nil, err
+		}
+		gRPCServer = grpc.NewServer(grpc.Creds(tlsCred))
+	} else {
+		gRPCServer = grpc.NewServer()
+	}
+	lis, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:%d", conf.Port))
+	if err != nil {
+		xdsServerLog.Errorf("xDS server failed to listen on %s: %v", fmt.Sprintf(":%d", conf.Port), err)
+		return nil, err
+	}
+	xdsServerLog.Infof("%s xDS server listens on %s", time.Now().String(), lis.Addr().String())
+	discovery.RegisterAggregatedDiscoveryServiceServer(gRPCServer, server)
+	snapshotCache.SetSnapshot("", cache.Snapshot{
+		Listeners: cache.Resources{Version: time.Now().String(), Items: map[string]cache.Resource{
+			"backend": ls.makeListener()}}})
+	go func() {
+		_ = gRPCServer.Serve(lis)
+	}()
+	return gRPCServer, nil
+}
+
+type XDSCallbacks struct {
+	numStream         int
+	numTokenReceived  int
+	callbackError     bool
+	lastReceivedToken string
+	mutex             sync.RWMutex
+	expectedToken     string
+	t                 *testing.T
+}
+
+func CreateXdsCallback(t *testing.T) *XDSCallbacks {
+	return &XDSCallbacks{t: t}
+}
+
+func (c *XDSCallbacks) SetCallbackError(setErr bool) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.callbackError = setErr
+}
+
+func (c *XDSCallbacks) SetExpectedToken(expected string) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.expectedToken = expected
+}
+
+func (c *XDSCallbacks) ExpectedToken() string {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	return c.expectedToken
+}
+
+func (c *XDSCallbacks) NumStream() int {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	return c.numStream
+}
+
+func (c *XDSCallbacks) NumTokenReceived() int {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	return c.numTokenReceived
+}
+
+func (c *XDSCallbacks) OnStreamOpen(ctx context.Context, id int64, url string) error {
+	xdsServerLog.Infof("xDS stream (id: %d, url: %s) is open", id, url)
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.numStream++
+	if metadata, ok := metadata.FromIncomingContext(ctx); ok {
+		if h, ok := metadata[credentialTokenHeaderKey]; ok {
+			if len(h) != 1 {
+				c.t.Errorf("xDS stream (id: %d, url: %s) sends multiple tokens (%d)", id, url, len(h))
+			}
+			if h[0] != c.lastReceivedToken {
+				c.numTokenReceived++
+				c.lastReceivedToken = h[0]
+			}
+			if c.expectedToken != "" && strings.TrimPrefix(h[0], "Bearer ") != c.expectedToken {
+				c.t.Errorf("xDS stream (id: %d, url: %s) sent a token that does "+
+					"not match expected token (%s vs %s)", id, url, h[0], c.expectedToken)
+			} else {
+				c.t.Logf("xDS stream (id: %d, url: %s) has valid token: %v", id, url, h[0])
+			}
+		} else {
+			c.t.Errorf("XDS stream (id: %d, url: %s) does not have token in metadata %+v",
```

Comment:
I don't think you can fail a test from a goroutine

Reply:
I think this works. I insert a c.t.Errorf("dummy error") before ```return nil``` here and the test fails with that error.
From https://golang.org/pkg/testing/#T, it says 
```A test ends when its Test function returns or calls any of the methods FailNow, Fatal, Fatalf, SkipNow, Skip, or Skipf. Those methods, as well as the Parallel method, must be called only from the goroutine running the Test function. The other reporting methods, such as the variations of Log and Error, may be called simultaneously from multiple goroutines.```